cmd/sovereignnode/dataCodec/dataCodec.go

```diff
@@ -0,0 +1,491 @@
+				Value: abi.EnumValue{Discriminant: uint8(tokenData.TokenType)},
+			},
+			{
+				Name:  "amount",
+				Value: abi.BigIntValue{Value: tokenData.Amount},
+			},
+			{
+				Name:  "frozen",
+				Value: abi.BoolValue{Value: tokenData.Frozen},
+			},
+			{
+				Name:  "hash",
+				Value: abi.BytesValue{Value: tokenData.Hash},
+			},
+			{
+				Name:  "name",
+				Value: abi.BytesValue{Value: tokenData.Name},
+			},
+			{
+				Name:  "attributes",
+				Value: abi.BytesValue{Value: tokenData.Attributes},
+			},
+			{
+				Name:  "creator",
+				Value: abi.AddressValue{Value: tokenData.Creator},
+			},
+			{
+				Name:  "royalties",
+				Value: abi.BigIntValue{Value: tokenData.Royalties},
+			},
+			{
+				Name: "uris",
+				Value: abi.InputListValue{
+					Items: tokenUris,
+				},
+			},
+		},
+	}
+
+	encodedTokenData, err := dc.serializer.Serialize([]any{tokenDataStruct})
+	if err != nil {
+		return nil, err
+	}
+
+	return hex.DecodeString(encodedTokenData)
+}
+
+// DeserializeTokenData will deserialize bytes to an esdt token data
+func (dc *dataCodec) DeserializeTokenData(data []byte) (*sovereign.EsdtTokenData, error) {
+	if len(data) == 0 {
+		return nil, errEmptyTokenData
+	}
+
+	enum := &abi.EnumValue{}
+	amount := &abi.BigIntValue{}
+	frozen := &abi.BoolValue{}
+	hash := &abi.BytesValue{}
+	name := &abi.BytesValue{}
+	attributes := &abi.BytesValue{}
+	creator := &abi.AddressValue{}
+	royalties := &abi.BigIntValue{}
+	uris := &abi.OutputListValue{
+		ItemCreator: func() any { return &abi.BytesValue{} },
+	}
+
+	tokenDataStruct := abi.StructValue{
+		Fields: []abi.Field{
+			{
+				Name:  "token_type",
+				Value: enum,
+			},
+			{
+				Name:  "amount",
+				Value: amount,
+			},
+			{
+				Name:  "frozen",
+				Value: frozen,
+			},
+			{
+				Name:  "hash",
+				Value: hash,
+			},
+			{
+				Name:  "name",
+				Value: name,
+			},
+			{
+				Name:  "attributes",
+				Value: attributes,
+			},
+			{
+				Name:  "creator",
+				Value: creator,
+			},
+			{
+				Name:  "royalties",
+				Value: royalties,
+			},
+			{
+				Name:  "uris",
+				Value: uris,
+			},
+		},
+	}
+
+	err := dc.serializer.Deserialize(hex.EncodeToString(data), []any{&tokenDataStruct})
+	if err != nil {
+		return nil, err
+	}
+
+	urisArg := make([][]byte, 0)
+	for _, item := range uris.Items {
+		uri, ok := item.(*abi.BytesValue)
+		if !ok {
+			return nil, errTypeAssertion
+		}
+
+		urisArg = append(urisArg, uri.Value)
+	}
+
+	return &sovereign.EsdtTokenData{
+		TokenType:  core.ESDTType(enum.Discriminant),
+		Amount:     amount.Value,
+		Frozen:     frozen.Value,
+		Hash:       hash.Value,
+		Name:       name.Value,
+		Attributes: attributes.Value,
+		Creator:    creator.Value,
+		Royalties:  royalties.Value,
+		Uris:       urisArg,
+	}, nil
+}
+
+// GetTokenDataBytes will receive token properties, deserialize then into a digital token and return marshalled digital token
+func (dc *dataCodec) GetTokenDataBytes(tokenNonce []byte, tokenData []byte) ([]byte, error) {
```

Comment:
This should return `&esdt.ESDigitalToken`, not the marshalled object, **or** value.Bytes.
The logic inside `	if digitalToken.Type == uint32(core.Fungible) ` is at caller level(incomingOperationsProcessor)

This is also better in case you use different marshallers in different compomnents. It's also somehow hidden and really confusing why this func returns either a marshalled struct or a value bytes. This small comp should only be capable of demarshallizing  bytes into a struct.

Reply:
Actually, You already have and use inside `.DeserializeTokenData`, so this func **should be completely deleted**.
This func should be defined inside `incomingEvents/DataProcessor`